src/Network/Ethereum/Web3/Encoding/Event.hs

```diff
@@ -0,0 +1,242 @@
+    as' <- arrayParser as
+    return $ I a' :* as'
+
+instance ArrayParser (NP f as) => ArrayParser (SOP f '[as]) where
+  arrayParser = fmap (SOP . Z) . arrayParser
+
+genericArrayParser :: ( Generic a
+                      , Rep a ~ rep
+                      , ArrayParser rep
+                      )
+                    => [T.Text]
+                    -> Maybe a
+genericArrayParser = fmap to . arrayParser
+
+
+--------------------------------------------------------------------------------
+-- Event Parsing
+--------------------------------------------------------------------------------
+
+data Event i ni = Event i ni
+
+parseChange :: ( Generic i
+               , Show i
+               , Rep i ~ irep
+               , ArrayParser irep
+               , Generic ni
+               , Show ni
+               , Rep ni ~ nirep
+               , GenericABIDecode nirep
+               )
+             => Change
+             -> Bool -- is anonymous event
+             -> Maybe (Event i ni)
+parseChange change isAnonymous = do
+    i <- genericArrayParser topics
+    ni <- genericFromData data_
+    return $ Event i ni
+  where
+    strip0x hx = if T.take 2 hx == "0x" then T.drop 2 hx else hx
+    topics = map strip0x $
+      if isAnonymous
+        then changeTopics change
+        else tail $ changeTopics change
+    data_ = strip0x $ changeData change
+
+class IndexedEvent i ni e | e -> i ni where
+  isAnonymous :: Proxy e -> Bool
+
+class CombineChange i ni e | e -> i ni where
+  combineChange :: i -> ni -> e
+
+instance ( Generic i
+         , Rep i ~ irep
+         , Generic ni
+         , Rep ni ~ nirep
+         , Generic e
+         , Rep e ~ erep
+         , HListRep irep hli
+         , HListRep nirep hlni
+         , MergeIndexedArguments hli hlni
+         , MergeIndexedArguments' hli hlni ~ hle
+         , HListRep erep hle
+         , IndexedEvent i ni e
+         ) => CombineChange i ni e where
+  combineChange i ni =
+    let hli = toHList . from $ i
+        hlni = toHList . from $ ni
+        hle = mergeIndexedArguments hli hlni
+    in to . fromHList $ hle
+
+class DecodeEvent i ni e | e -> i ni where
+  decodeEvent :: Change -> Maybe e
+
+
+instance ( IndexedEvent i ni e
+         , Generic i
+         , Rep i ~ SOP I '[hli]
+         , Generic ni
+         , Rep ni ~ SOP I '[hlni]
+         , Generic e
+         , Rep e ~ SOP I '[hle]
+         , CombineChange i ni e
+         , GenericABIDecode (SOP I '[hlni])
+         , ArrayParser (SOP I '[hli])
+         , Show i, Show ni
+         ) => DecodeEvent i ni e where
+  decodeEvent change = do
+      let anonymous = isAnonymous (Proxy :: Proxy e)
+      (Event i ni :: Event i ni) <- parseChange change anonymous
+      return $ combineChange i ni
+
+--------------------------------------------------------------------------------
+-- Event Parsing Internals
+--------------------------------------------------------------------------------
+
+data HList :: [*] -> * where
```

Comment:
So, may be this type and sorting classes can be moved to separate module from `Event` module?